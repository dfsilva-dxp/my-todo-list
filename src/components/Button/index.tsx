import * as S from "./styles";

import { IButtonProps } from "./types";

const Button = ({ children, icon }: IButtonProps) => {
  return (
    <S.ButtonContent>
      {!!children && children}
      {!!icon && icon}
    </S.ButtonContent>
  );
};

export default Button;
