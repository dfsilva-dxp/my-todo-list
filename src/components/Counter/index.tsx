import * as S from "./styles";

import { ICounterProps } from "./types";

const Counter = ({ value }: ICounterProps) => {
  return (
    <S.Wrapper role="badge" aria-label="counter">
      {value}
    </S.Wrapper>
  );
};

export default Counter;
