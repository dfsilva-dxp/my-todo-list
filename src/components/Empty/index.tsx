import clipboardSVG from "@assets/clipboard.svg";

import * as S from "./styles";

const Empty = () => {
  return (
    <S.Wrapper>
      <img
        src={clipboardSVG}
        alt="clipboard icon"
        title="clipboard icon"
        aria-label="clipboard icon"
      />

      <S.TextEmptyWrapper>
        <h4>Você ainda não tem tarefas cadastradas</h4>
        <p>Crie tarefas e organize seus itens a fazer</p>
      </S.TextEmptyWrapper>
    </S.Wrapper>
  );
};

export default Empty;
