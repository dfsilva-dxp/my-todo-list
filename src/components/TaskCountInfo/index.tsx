import * as S from "./styles";

import { ITaskCountInfoProps } from "./types";

const TaskCountInfo = ({
  children,
  badge,
  variant = "created"
}: ITaskCountInfoProps) => {
  return (
    <S.Wrapper variant={variant} role="info" aria-label="count info">
      <p>{children}</p>
      {badge}
    </S.Wrapper>
  );
};

export default TaskCountInfo;
