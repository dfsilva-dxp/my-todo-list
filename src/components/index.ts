export { default as Button } from "./Button";
export { default as Checkbox } from "./Checkbox";
export { default as Container } from "./Container";
export { default as Counter } from "./Counter";
export { default as Empty } from "./Empty";
export { default as Form } from "./Form";
export { default as Header } from "./Header";
export { default as Input } from "./Input";
export { default as Logo } from "./Logo";
export { default as Task } from "./Task";
export { default as TaskCountInfo } from "./TaskCountInfo";
export { default as TaskList } from "./TaskList";
