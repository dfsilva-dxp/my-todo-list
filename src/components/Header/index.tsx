import { Logo } from "@/components";

import * as S from "./styles";

import logoSVG from "@/assets/logo.svg";

const Header = () => {
  return (
    <S.Wrapper role="banner">
      <Logo
        src={logoSVG}
        title="Logo Todo App"
        alt="Logo Todo App"
        aria-label="logo"
      />
    </S.Wrapper>
  );
};

export default Header;
