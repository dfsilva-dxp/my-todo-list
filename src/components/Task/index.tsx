import { useState } from "react";
import { Trash } from "phosphor-react";

import { Button, Checkbox } from "@/components";

import useTaskContext from "@/hooks/useTasksContext";

import * as S from "./styles";

import { ITaskProps } from "./types";

const Task = ({ description, checked, taskId }: ITaskProps) => {
  const [isChecked, setIsChecked] = useState(checked);

  const { updateTask } = useTaskContext();

  const onCheck = (status: boolean) => {
    setIsChecked(status);
    updateTask(taskId, status);
  };

  return (
    <S.Wrapper role="listitem" aria-label="task card">
      <Checkbox isChecked={isChecked} onCheck={onCheck} />

      <S.TaskDescription status={isChecked ? "done" : "created"}>
        {description}
      </S.TaskDescription>

      <Button
        type="button"
        minimal
        size="small"
        icon={<Trash size={20} weight="bold" />}
      />
    </S.Wrapper>
  );
};

export default Task;
