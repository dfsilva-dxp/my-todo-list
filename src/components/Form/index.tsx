import { FormEvent, useState } from "react";
import { PlusCircle } from "phosphor-react";

import { Button, TextField } from "@/components";

import useTaskContext from "@/hooks/useTasksContext";

import * as S from "./styles";

import { ITask } from "@components/Task/types";

const Form = () => {
  const [inputValue, setInputValue] = useState("");

  const { createTask } = useTaskContext();

  const onInputChange = (value: string) => {
    setInputValue(value);
  };

  const handleSubmit = (evt: FormEvent<HTMLFormElement>) => {
    evt.preventDefault();
    const id = Math.floor(Date.now() * Math.random());

    const data: ITask = {
      id: String(id),
      description: inputValue,
      status: "created",
      checked: false,
      createdAt: new Date()
    };

    createTask(data);
    setInputValue("");
  };

  return (
    <S.Wrapper onSubmit={handleSubmit}>
      <TextField
        placeholder="Adicione uma nova tarefa"
        initialValue={inputValue}
        onInputValue={onInputChange}
      />
      <Button
        type="submit"
        size="small"
        icon={<PlusCircle size={20} weight="bold" />}
      >
        Criar
      </Button>
    </S.Wrapper>
  );
};

export default Form;
