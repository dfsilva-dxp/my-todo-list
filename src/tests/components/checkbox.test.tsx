import { renderWithTheme } from "@/utils/tests/renderWithTheme";
import { act, screen, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import { Checkbox } from "@/components";

describe("<Checkbox />", () => {
  it("should render with label", () => {
    renderWithTheme(<Checkbox label="checkbox label" labelFor="check" />);

    expect(screen.getByRole("checkbox")).toBeInTheDocument();

    expect(screen.getByLabelText(/checkbox label/i)).toBeInTheDocument();
    expect(screen.getByText(/checkbox label/i)).toHaveAttribute("for", "check");
  });

  it("should render without label", () => {
    renderWithTheme(<Checkbox />);

    expect(screen.queryByLabelText("checkbox")).not.toBeInTheDocument();
  });

  it("should dispatch onCkeck(true) when status change", async () => {
    const onCheck = jest.fn();

    renderWithTheme(
      <Checkbox label="checkbox label" labelFor="check" onCheck={onCheck} />
    );

    expect(onCheck).not.toHaveBeenCalled();

    act(() => {
      userEvent.click(screen.getByRole("checkbox"));
    });

    await waitFor(() => {
      expect(onCheck).toHaveBeenCalledTimes(1);
    });

    expect(onCheck).toHaveBeenCalledWith(true);
  });

  it("should dispatch onCkeck(false) when status change", async () => {
    const onCheck = jest.fn();

    renderWithTheme(
      <Checkbox
        label="checkbox label"
        labelFor="check"
        onCheck={onCheck}
        isChecked
      />
    );

    act(() => {
      userEvent.click(screen.getByRole("checkbox"));
    });

    await waitFor(() => {
      expect(onCheck).toHaveBeenCalledTimes(1);
    });

    expect(onCheck).toHaveBeenCalledWith(false);
  });

  it("should be accessible with tab", async () => {
    const { container } = renderWithTheme(
      <Checkbox label="Checkbox" labelFor="Checkbox" />
    );

    expect(document.body).toHaveFocus();

    await userEvent.tab();

    expect(screen.getByLabelText(/checkbox/i)).toHaveFocus();

    expect(container.firstChild).toMatchSnapshot();
  });
});
