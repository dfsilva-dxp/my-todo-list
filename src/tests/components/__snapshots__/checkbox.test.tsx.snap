// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Checkbox /> should be accessible with tab 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: 1rem;
}

.c1 {
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  width: 1.5rem;
  height: 1.5rem;
  border: 0.2rem solid #4EA8DE;
  border-radius: 100%;
  -webkit-transition: background border 0.15s ease-in-out;
  transition: background border 0.15s ease-in-out;
  position: relative;
  outline: none;
}

.c1:before {
  content: "";
  width: 0.3rem;
  height: 0.6rem;
  border: 0.15rem solid #F2F2F2;
  border-top: 0;
  border-left: 0;
  -webkit-transform: rotate(45deg);
  -ms-transform: rotate(45deg);
  transform: rotate(45deg);
  position: absolute;
  top: 0.1rem;
  right: 0.3rem;
  opacity: 0;
  -webkit-transition: 0.15s ease-in-out;
  transition: 0.15s ease-in-out;
}

.c1:focus {
  border-color: #5E60CE;
}

.c1:checked {
  border-color: #5E60CE;
  background: #5E60CE;
}

.c1:checked:before {
  opacity: 1;
}

.c2 {
  cursor: pointer;
  font-size: 0.875rem;
  font-weight: 700;
  line-height: 1.8rem;
}

<div
  class="c0"
>
  <input
    class="c1"
    id="Checkbox"
    role="checkbox"
    type="checkbox"
    value=""
  />
  <label
    class="c2"
    for="Checkbox"
  >
    Checkbox
  </label>
</div>
`;
